### ISSUE TITLE
switch build system

### ISSUE THREAD

**Issue by stsp** [stsp/dosemu2#393]
I've got entirely frustrated with continuous
tweaking of our hand-written makefiles and
want something else to do. :)
Every new small feature like deb integration
or separate build dir is a major break-through
for us and costs weeks of work, while other
projects just have this all for free.
There is our custom build extension called
mkpluginhooks, that allows to easily add new
pieces of code to the project without modifying
the main configure.ac and the rest of the
code-base. But I spent a week minimizing its
role in a build process, and now it is needed
mostly just to dynamically extend the list of
the build directories.

So any new system we chose, should have
that capability: the pieces of code can be
easily added to src/plugins/* and it should be
able to pick that up itself. No other special
requirements I can think of, do exist. Of course
we can completely remove the plugin system,
but IMHO that would be upsetting, so I hope
for the build system that can support that.

I am not familiar with any build system other
than automake unfortunately, and I don't think
we need to switch to automake. Or do we?
So it will require some time before I can do
the choice willingly.
For now I made a quick look into a modern
build systems, and it seems that we have to
choose between scons, cmake and premake.
Looking at the syntax and different examples,
it looks like pre-make is a best choice: lua
syntax, very easy to read those few examples
I've seen. scons looks too complex from the
first glance, and cmake's syntax looks like
entire crap (google reveals that it is not only
my opinion, but it may be that syntax is the
only complain it gets, and works quite well
otherwise).

What build system do we need?
Please share your experience. :)

**Comment by stsp** [comment]
Or maybe indeed just switch to automake
to minimize an efforts...

**Comment by stsp** [comment]
https://stackoverflow.com/a/18270096/5515547
Nice comparison of build systems.

**Comment by stsp** [comment]
There seem to be a very new and promising
build system for gnome project:
http://mesonbuild.com/
Good syntax (DSL, but much better than cmake's
one to my taste), very informative, short and simple
docs, lots of configuration features we got used to
in autoconf.
But it does not allow you to use anything else but
its DSL, unlike eg waf, where python is at your
disposal. This means things like mkpluginhooks
(and any other external scripts) are not possible.

So my current plan is to eradicate mkpluginhooks,
default-configure and the rest of an external
complexity, then start looking into this mesonbuild
thing.

**Comment by tkchia** [comment]
Sorry for the late reply (especially after I filed the initial `$builddir` != `$srcdir` patches).

For what it is worth, I think the `Makefile` setup coupled with use of `VPATH` works fine when the targets and dependencies are explicitly specified (since the `VPATH` mechanism sort of hides the fact that there are two directories involved).  The problems mainly arise when the makefiles use wildcards --- as in the `bdftopcf` conversions in `src/arch/linux/Makefile.main` --- since the wildcards force us to explicitly specify which directory to search for files in.

Regarding Meson, I have not actually tried it, so hopefully it addresses this wildcards issue in some way, or it has other advantages. :-)

**Comment by stsp** [comment]
Please just take a look (from the git log) at how
many efforts have I spent bringing your initial
patches alive. :)
Its not something that can't be done, but it is
too many work and is bug-prone. I suppose
automake and other build systems also use
vpath, so its not something we are going to
miss if we switch: it will still be there, but w/o
so much of the hand-written complexity.
vpath is very good when you write your own
build system, which is what we unfortunately
do quite often. And it is very good for the
automatic build-system generators.

```
$ find . -name 'Make*' -printf '%s + ' |d 's/$/0/' | xargs expr
81430
```

So we are maintaining ~80K of a hand-written
makefile code... The features like deb and rpm builds,
`make uninstall`, separate dir builds etc - these all
are for free with build systems and a major break-through
for us. Not good. Also think about cygwin or bsd ports.

**Comment by stuaxo** [comment]
Another possible +1 for Meson is that build times may be significantly improved - this is certainly the case on gstreamer [1] and epoxy [2] (GNOME opengl bindings).

[1] http://blog.nirbheek.in/2016/05/gstreamer-and-meson-new-hope.html

[2] https://www.phoronix.com/scan.php?page=news_item&px=Meson-Epoxy-Speed

**Comment by stsp** [comment]
Personally, I do not care about the build times. :)
Note that full `autogen.sh` + configure probably takes
more than the build time itself, so if we are to reduce
build times, we should first blame autoconf & friends,
not gnu make.
BUT, when I look at the horrid of `make -d` output,
I am starting to think that being so ineffective is bad.

**Comment by waruqi** [comment]
You can also try another modern build system based lua.  [xmake](http://xmake.io)

**Comment by stsp** [comment]
It doesn't say it is a build system,
like automake or meson is. It says only:
`A make-like build utility based on Lua`.
What we certainly don't need is yet another
make.
Please correct me if I misunderstand what
xmake is.

**Comment by waruqi** [comment]
@stsp It's also a build system, like cmake, premake,scons and It can build project directly or generate IDE project files (vcproj, makefile, ..)

You can see an example about the project decription file [xmake.lua](https://github.com/waruqi/libuv/blob/master/xmake.lua)

**Comment by stsp** [comment]
Ok, maybe they are incredibly humble, and do
not manifest their features to the mere mortals. :)
So what they list, is:
```
Support features

    Tasks
    Macros
    Actions
    Options
    Plugins
    Templates
```

I am not sure in what of the above the
"configuration system" hides. In the example
you pointed, the only check I can see is
`is_plat()`. Is there an example that also
checks libraries and compiler features?
Trying to google `xmake vs cmake` gives
nothing. Surely the xmake guys are too
humble.
Meson attracts by the mere fact that it
is trying to advertise itself. Its documentation
is written in a way you will know you need
it, even if you don't. :) If xmake guys do not
want to attract users, then I wonder why
should I be attracted.

**Comment by waruqi** [comment]
@stsp I'm sorry, this is only because my English is not good, in the document did not express clearly. :()

About compiler features, you can see [compiler features detection](http://tboox.org/2017/07/29/new-features-v2.1.5/).

**Comment by stsp** [comment]
Hmm, okey, thanks.
Is there a design goals description,
comparison to other build systems,
and the list of project that use it?

**Comment by waruqi** [comment]
@stsp It is designed to make it easier to maintain and build project. It provides simple lua syntax, not complex DSL and flexible, extensible custom scripts and plugins.

We can also intelligently scan the source directory to automatically build all source files without any make-like file. 
 
There are [some projects](https://github.com/tboox/xmake/wiki/xmake-projects) using xmake, but not many.

**Comment by stsp** [comment]
Meson docs state that providing the DSL is an advantage,
as with the turing-complete scripting the one can produce
unportable recipes.
In any case, while xmake's docs are not even nearly
as rich, I probably need to try this out and see whether
it suits or not. In fact, before finding meson, I was
actually searching for something lua-style...
And this is exactly the situation where the in-depth
comparison articles can help so that I don't have to
do that myself. :)

**Comment by waruqi** [comment]
Sorry, xmake does not have too many [English documents](http://xmake.io/#/home), I am trying to translate them. 😄

**Comment by stsp** [comment]
> I am trying to translate them. :smile:

Maybe you can also translate 1-2 dosemu
plugins to xmake as an example? :smile:

**Comment by waruqi** [comment]
If your uses xmake to maintain and build dosemu, I will add it's `xmake.lua` to the document as an example. 😄

**Comment by stsp** [comment]
dosemu consists of many plugins, and most of
`configure.ac` files just look like this:
```
dnl Process this file with autoconf to produce a configure script.
AC_INIT
AC_CONFIG_FILES([Makefile.conf])
AM_PATH_ALSA(0,,
    AC_MSG_ERROR([alsa development libs not found]))
AC_SUBST(ALSA_CFLAGS)
AC_SUBST(ALSA_LIBS)
AC_OUTPUT
```
and same to corresponding Makefile, because the
plugins are usually implemented with a single small
C file.

It should be absolutely trivial to convert 1-2 of
them as an example. You don't have to do that,
but it should really be simple if the build system
is mature enough.

**Comment by waruqi** [comment]
Is like this?

root xmake.lua:

```lua
set_project("dosemu")

-- some targets

includes("plugin/alsa", "plugin/libao")
```

`plugin/alsa`:
```lua
target("plugin_alsa")
    set_kind("shared")
    add_files("*.c")
```

`plugin/libao`:
```lua
target("plugin_libao")
    set_kind("shared")
    add_files("*.c")
```

**Comment by waruqi** [comment]
Or add plugins dynamically

```lua
for _, plugindir in ipairs(os.dirs("plugin/*")) do
    target("plugin_" .. path.basename(plugindir))
        set_kind("shared")
        add_files(path.join(plugindir, "**.c"))
end
```

**Comment by stsp** [comment]
I'd like to do this at first without modifying root.
I guess this is not a problem if every plugin is a
separate project?

**Comment by stsp** [comment]
I mean, in this case we can actually commit
the change in a separate branch, that will have
just 1-2 plugins building with xmake, leaving
everything else alone.

**Comment by waruqi** [comment]
> I'd like to do this at first without modifying root.
I guess this is not a problem if every plugin is a
separate project?

Yes, you can write a separate `xmake.lua` for each plugin and need not modify the root `xmake.lua`.

> I mean, in this case we can actually commit
the change in a separate branch, that will have
just 1-2 plugins building with xmake, leaving
everything else alone.

You can only add `xmake.lua` to 1-2 plugins and add a root `xmake.lua` to include them and call xmake to build them in other build system. e.g makefile, ..

**Comment by stsp** [comment]
> and add a root xmake.lua

Why should I do that?
is it not possible to build them independently
without root lua?

**Comment by stsp** [comment]
All plugins are just a self-contained solibs,
surely they can be separately built w/o centralized
script?

**Comment by waruqi** [comment]
Just add a line `includes("plugin/**/xmake.lua")` to the root script. If you don't want to add root script, you need pass every plugin script directories to xmake for building them directly.

For example:

```console
$ xmake -P ./plugin/libao
$ xmake -P ./plugin/xxx
```

**Comment by stsp** [comment]
> as argument to xmake for building them directly.

Yes, that's what we do now.
We go to all plugin dirs and do autoconf+configure
there. No root configure script that would
include them, and I'd like that logic to stay.

**Comment by waruqi** [comment]
Yes, you can only run the following commands to build the given plugins without root configure.

```console
$ xmake -P ./plugin/xxx
```

Or 

```console
$ xmake -F plugindir/xxx/xmake.lua
```

**Comment by stsp** [comment]
What features will we get?
Currently we implemented the out-of-tree
builds and integration with package managers:
deb and rpm. We also implement a lot of
useful targets like "clean", "dist" (for tar-ball),
"changelog" etc. With what pre-defined targets
will it provide us? Because what I don't want is
to have to re-implement them in a different
build system - I want them to be there from
the very beginning, so that I can only remove
our current build-system code, and not to write
anything for a replacement. :) Of course I
admit that the configure.ac will have to be
re-implemented, but the logic we have now
in makefiles should not be coded by us.

**Comment by stsp** [comment]
> Remote package management: I'm trying to develop.

This doesn't seem to list deb or rpm.
But for dosemu2 this probably doesn't matter
because I already wrote the package-specific
build recipes (spec for rpmbuild and controls/rules
for debuild), so the help from the build system
to generate them is no longer needed.

> Currently not supported

Even the out-of-tree build?

> xmake is still in the development stage, of course, if you don't
> think it is suitable for your project, you can also try some other build systems.

Ideally I'd like to try out a few by converting
a couple of plugins, then see. So while I try
to find the reason to _not_ try xmake (as I
don't have enough time), I may as well try it. :)

> Perhaps to facilitate the development and debugging of programs with the help of the IDE. : )

If you generate the ide-specific project files
then yes. But most of them generate the
plain makefiles. Maybe there is a reason of
doing so, but I am just not aware of it.
Reading about ninja, it seems to be trying
to build the common low-level language for
the different build systems, which is, as they
say, very easy to generate but very difficult
to write by hands. Maybe that makes sense.

**Comment by waruqi** [comment]
> Even the out-of-tree build?
It is supported, only not support to generate deb or rpm now.

> Ideally I'd like to try out a few by converting
a couple of plugins, then see. So while I try
to find the reason to not try xmake (as I
don't have enough time), I may as well try it. :)

I see. 😃

**Comment by stsp** [comment]
> It is supported, only do not support generating deb or rpm now.

That looks good because only out-of-tree builds
are important for us. I don't think any build system
can generate the _good_ rpm and deb packages,
it can likely only generate the very basic ones.
We already have the very customized recipes
for deb and rpm, so no problem with them.

I wonder if the build system could help with the
git integration, so that we finally stop having this:
https://github.com/stsp/dosemu2/issues/356
https://github.com/stsp/dosemu2/issues/326#issuecomment-304318951
https://github.com/stsp/dosemu2/issues/324
https://github.com/stsp/dosemu2/issues/411
All I need is to rebuild the version file when
the git head changes...

**Comment by stsp** [comment]
> That looks good because only out-of-tree builds
> are important

... and DESTDIR. Certainly you have DESTDIR support?

**Comment by waruqi** [comment]
> I wonder if the build system could help with the
git integration, so that we finally stop having this:
#356
#326 (comment)
#324
#411
All I need is to rebuild the version file when
the git head changes...

You can import git module to get git tags and rewrite version file before building project.

```lua
target("test")
    before_build(function (target)
         import("devel.git")
         local version = git.tags()[1]
         if version then
              io.gsub(versionfile, "[match_pattern]", version)  -- rewrite version 
         end
    end)
```

Or detect git version to show some tips for low version.

```lua
target("test")
    before_build(function (target)
         import("lib.detect.find_tool")
         local gitinfo = find_tool("git", {version = true})
         assert(gitinfo.version > "2.0.0", "the current version of git is too low!")
    end)
```

Or run some other git commands which the builtin git module has not provide:

```lua
os.run("git rev-parse --symbolic-full-name HEAD")
```

If the current git do not support some argument, you can use try-catch to handle error:

```lua
try
{ 
    function ()
        os.run("git rev-parse --symbolic-full-name HEAD")
    end,
    catch 
    {
         function (errors)
                raise("git rev-parse failed, %s", errors or "")
         end
    }
}
```

**Comment by waruqi** [comment]
> ... and DESTDIR. Certainly you have DESTDIR support?

like this?  

```console
$ xmake install -o destdir
```

**Comment by stsp** [comment]
> You can import git module to get git tags and rewrite version file before building project.

Version is written into a `.h` file which then
gets included. So the important thing is to
not rewrite it unless the version actually
changed, to not trigger the extra rebuilds.
For that I currently use the evil tricks like
the tmpfile created as follows:
`git log -1 --format=%cd --date=rfc | xargs -I {} touch --date={} $TSTAMP`
on which the generated header file then
depends. This also required to use the
`.LOW_RESOLUTION_TIME:` makefile
target. Previously I tried to depend directly
on the ref file in `.git`, but that was broken
with every new git version. Ideally the build
system should handle that itself. For example
by allowing to specify the timestamps explicitly,
so that the target gets re-built when its
modification time predates the timestamp.
Then I can get the timestamp of the last
commit from git log, and make the version
header to be re-built when the git log's
timestamp is newer.

> like this?

Not quite, the build system should take the
`DESTDIR` variable without any extra options,
because the package builders just set
DESTDIR and expect it to be honored,
if my understanding is correct.

**Comment by waruqi** [comment]
> Version is written into an .h file which then
gets included.

You can try 

```lua
set_version("2.1.6", {build = "%Y%m%d%H%M"})
set_config_header("config.h")
```

It will generate version macro to `config.h`, for example:

```
#define CONFIG_VERSION_MAJOR         2
#define CONFIG_VERSION_MINOR         1
#define CONFIG_VERSION_ALTER         6
#define CONFIG_VERSION_BUILD         201708272125
```

> So the important thing is to
not rewrite it unless the version actually
changed, to not trigger the extra rebuilds.
For that I currently use the evil tricks like
the tmpfile created as follows:
git log -1 --format=%cd --date=rfc | xargs -I {} touch --date={} $TSTAMP
on which the generated header file then
depends.

You can call `os.mtime("filepath")` to get the modification time of dependent file or get the previous version from the version.h to compare it. If the version has been changed, then we will update the new version to file.

> Not quite, the build system should take the
DESTDIR variable without any extra options,
because the package builders just set
DESTDIR and expect it to be honored,
if my understanding is correct

like this?  

```
set_targetdir("$(buildir)/xxx")
set_objectdir("$(buildir)/.objs")
set_headerdir("$(projectdir)/inc")
```

**Comment by stsp** [comment]
> You can call os.mtime("filepath") to get the modification time of dependent
> file or get the previous version from the version.h to compare it.

So you are proposing to again do this by hands.
Plus you don't say how should I trigger the rebuild
if I found that mtime is old enough. Perhaps the
explicit timestamp is not possible here as it is not
possible in `make`, so to trigger the rebuild, I'll
need to update the mtime of tmpfile and depend
on that tmpfile, as I do now with make.
But can the build system just provide the recipe
for this?

> like this?

So you mean the DESTDIR handling must be
again done by hands? No built-in support?

**Comment by waruqi** [comment]
> Plus you don't say how should I trigger the rebuild
if I found that mtime is old enough. Perhaps the
explicit timestamp is not possible here as it is not
possible in make, so to trigger the rebuild, I'll
need to update the mtime of tmpfile and depend
on that tmpfile, as I do now with make.
But can the build system just provide the recipe
for this?

You can call `task.run("rebuild", {target = "xxx"})` or `task.run("clean", {target = "xxx"})` to trigger 
  the rebuild.

```lua
target("test")
    before_build(function (target)
         import("core.base.task")
         if os.mtime("xxx") > ... then  -- need rebuild?
                task.run("clean", {target = target:name()}) -- clean all object files of this target to trigger the rebuild
         end
    end)
```

> So you mean the DESTDIR handling must be
again done by hands? No built-in support?

Isn't that a built-in support? 

```lua
set_targetdir("$(buildir)/xxx")
set_objectdir("$(buildir)/.objs")
set_headerdir("$(projectdir)/inc")
```

Or 

```lua
option("DESTDIR")
    set_default("/tmp/xxx")

target("test")
    set_targetdir("$(DESTDIR)")
```

The default output directory of target is `/tmp/xxx` and you can also modify it by `$xmake f --DESTDIR=/tmp/xxx2`

**Comment by stsp** [comment]
DESTDIR alters only "make install" so I don't
think I need to do `set_objectdir` etc - these are
unaffected. I suppose `set_targetdir("$(DESTDIR)")`
is what should be done.

> The default output directory of target is /tmp/xxx
> and you can also modify it by $xmake f --DESTDIR=/tmp/xxx2

But it is set by the package builders (debuild, rpmbuild)
via environment variable, not cmdline option.
https://www.gnu.org/software/automake/manual/html_node/DESTDIR.html
IMHO this should be supported by the build system
the same way automake does.

**Comment by stsp** [comment]
Also meson does the same:
https://github.com/mesonbuild/meson/blob/master/docs/markdown/Quick-guide.md
```
This is very similar to other build systems. The only difference is that the DESTDIR variable is passed as an environment variable rather than as an argument to ninja install.
```

**Comment by waruqi** [comment]
> But it is set by the package builders (debuild, rpmbuild)
via environment variable, not cmdline option.

You can get DESTDIR env and implement custom installation.

```lua
target("test")
    on_install(function (target)
        os.cp(target:targetfile(), "$(env DESTDIR)/lib/")   -- or path.join(os.getenv("DESTDIR"), "lib/")
        
        -- install headers
        -- ...
    end)
```

**Comment by stsp** [comment]
Oh c'mon, implementing "make install"
by hands, and manually taking care of
DESTDIR? But this is what we already
do with plain make...
Build systems should provide the
pre-defined targets like "make install"
that handle DESTDIR and other things
by themselves.
So I still don't quite understand the point
of xmake: should it be compared to make
(with many extensions, better language etc),
or should it be compared to the build systems
like autotools and meson?

**Comment by waruqi** [comment]
I have supported DESTDIR and PREFIX environment variables on master branch.

```console
   $ xmake install -o /usr/local
or $ export DESTDIR=/usr/local; xmake install 
or $ export INSTALLDIR=/usr/local; xmake install 
```

And we can also set PREFIX variable:

```console
   $ xmake install --prefix=local
or $ export PREFIX=local; xmake install 
```

**Comment by stsp** [comment]
Hmm. I cloned the repo and do `make build`.
It compiles something. If I do `make build` again,
it compiles everything again. And again, and
again... Wtf? :) I thought the build system
should build things just once, and then only
track what have changed.

**Comment by stsp** [comment]
I also can't find the resulting `xmake` binary...

**Comment by waruqi** [comment]
This command only compile the self code of xmake project. You can see [installation docs](http://xmake.io/#/home?id=installation) to install xmake binary.

Then we can create an example project and build it.

```console
xmake create -t 1 -P test
cd test
xmake
```

Or

```console
echo '#include <stdio.h>\nint main(){printf("hello\n");}' > main.c
xmake run
```

**Comment by stsp** [comment]
I don't understand what's going on.
I executed the script from the install docs,
it compiled something and installed xmake
to ~/.local/share/xmake (and to ~/.local/bin).
But I still can't find its binary in a source dir,
can't compile it manually, etc. Why would I
need some script that DLs, compiles and
installs? I want to do that myself.

**Comment by waruqi** [comment]
You can see the [usage video](https://asciinema.org/a/133693) 😄

**Comment by stsp** [comment]
No, I want to build and install xmake itself.
The video only shows usage. Why can't I
just clone the repo, compile and install?
Why "make build" always rebuilds everything?

**Comment by waruqi** [comment]
please see [compilation install doc](http://xmake.io/#/home?id=compilation) If you want to compile and install.

Or run 

```console
$ make build
$ sudo make install
```

**Comment by stsp** [comment]
But why `make build` rebuilds everything
without checking what was already built?
When I run `make` in other projects, on
the second run it just says "nothing to be
done for all". Is rebuilding everything is
the best of what the build system could do? :)

How can I only compile the xmake binary
and do not install it? Whatever I do I can't
find the built xmake binary in the source tree.

**Comment by waruqi** [comment]
The binary file: `./core/src/demo/demo.b`

Because we use the old build system to build xmake binary now. It does not yet support header file dependency compilation now. So we need rebuild xmake binary when installing and developing.

We do not modify the c codes in `./core` often and modify lua scripts primarily, so this not a problem.

We can use xmake to build self If we have installed xmake.

```console
$ cd core
$ xmake
```

The binary file: `./core/build/xmake`

**Comment by stsp** [comment]
> Because we use the old build system to build xmake binary now.
> It does not yet support header file dependency compilation now.

Understood, but I find it funny that the build system
has so many problems with the build system. :)

> We can use xmake to build self If we have installed xmake.

Yes, that works better but still not perfect:
```
$ xmake
[80%]: archiving.release libxmake.a
[100%]: linking.release xmake
$ xmake
[80%]: archiving.release libxmake.a
[100%]: linking.release xmake
```

So it still does the (20% of) unneeded work.

**Comment by waruqi** [comment]
> So it still does the (20% of) unneeded work.

Yes, I have fixed this problem on master branch.  You can fetch and try it again.

**Comment by stsp** [comment]
Yes, seems to work better now, thanks.

**Comment by stsp** [comment]
```
$ time xmake

real	0m1,654s
user	0m0,723s
sys	0m0,901s
```

Any idea why it spends ~2sec on nothing,
even when the data is already in the cache?
This is not really a complain as I have not
compared it to other build systems. But
subjectively that looks too much.

**Comment by waruqi** [comment]
Because xmake need check if the dependent header files, compilation flags are changed. It will detect if some flags in `xmake.lua` are supported for the current compiler when making compilation flags.

The current detection of flags is time-consuming, although we have done some optimizations.

```console
$ xmake -v
checking for the flags(xcrun -sdk macosx clang) -Qunused-arguments ... ok
checking for the flags(xcrun -sdk macosx clang) -Qunused-arguments ... ok
checking for the flags(xcrun -sdk macosx clang) -Wno-error=deprecated-declarations ... ok
checking for the flags(xcrun -sdk macosx clang) -fno-strict-aliasing ... ok
checking for the flags(xcrun -sdk macosx clang) -Wno-error=nullability-completeness ... ok
checking for the flags(xcrun -sdk macosx clang) -fomit-frame-pointer ... ok
checking for the flags(xcrun -sdk macosx clang) -O3 ... ok
checking for the flags(xcrun -sdk macosx clang) -Wno-nullability-completeness ... ok
checking for the flags(xcrun -sdk macosx clang) -Os ... ok
checking for the flags(xcrun -sdk macosx clang) -Os ... ok
checking for the flags(xcrun -sdk macosx clang++) -Qunused-arguments ... ok
checking for the flags(xcrun -sdk macosx clang++) -Wl,-rpath=@loader_path ... no
checking for the flags(xcrun -sdk macosx clang++) -Xlinker -rpath -Xlinker @loader_path ... ok
checking for the flags(xcrun -sdk macosx clang++) -Wl,-rpath=@loader_path/release/x86_64 ... no
checking for the flags(xcrun -sdk macosx clang++) -Xlinker -rpath -Xlinker @loader_path/release/x86_64 ... ok
checking for the flags(xcrun -sdk macosx clang++) -all_load ... ok
checking for the flags(xcrun -sdk macosx clang++) -pagezero_size 10000 ... ok
checking for the flags(xcrun -sdk macosx clang++) -image_base 100000000 ... ok
```

**Comment by waruqi** [comment]
I optimized it again. 

```console 
$ time xmake
real    0m0.532s 
user    0m0.368s 
sys     0m0.139s
```

**Comment by stsp** [comment]
I am getting lots of errors like this:
```
checking for the libm_atan2f ... ok
/home/stas/.local/share/xmake/core/base/os.lua:653: /home/stas/.local/share/xmake/core/base/os.lua:653: /tmp/.xmake500/_1AE2941B00682AAD5C99BABE55E2419B.c: В функции «main»:
/tmp/.xmake500/_1AE2941B00682AAD5C99BABE55E2419B.c:8:40: ошибка: «pwrite64» undeclared (first use in this function); did you mean «ppwrite64»?
     volatile void* ppwrite64 = (void*)&pwrite64;;
                                        ^~~~~~~~
                                        ppwrite64
stack traceback:
    [C]: in function 'error'
    [/home/stas/.local/share/xmake/core/base/os.lua:653]: in function 'raise'
    [/home/stas/.local/share/xmake/core/sandbox/modules/os.lua:330]: in function 'iorunv'
    [/home/stas/.local/share/xmake/modules/core/tools/gcc.lua:415]:

stack traceback:
    [C]: in function 'error'
    [/home/stas/.local/share/xmake/core/base/os.lua:653]: in function 'raise'
    [...xmake/core/sandbox/modules/import/core/tool/compiler.lua:111]: in function 'compile'
    [...ocal/share/xmake/modules/lib/detect/check_cxsnippets.lua:184]:
```

Although it builds.

**Comment by waruqi** [comment]
This is not an error, just is verbose backtrace information when detecting a c api. `$ xmake -v`. 

You can remove the `-v` argument to  do not see the verbose output.

**Comment by stsp** [comment]
I just wanted to see the "checking XXX"
messages which do not seem to appear
without `-v`. Perhaps you can implement
`-vv` that will also enable the lua verbose
messages.
In any case, I confirm that the 2 second
delay is gone! So at that point I see no
compelling reason why xmake should not
be used with dosemu2. :)
Likely I need to try it out to see one...

**Comment by stsp** [comment]
By the way, you seem to be taking the names
that were already taken.

xmake is taken:
https://www.codeproject.com/Articles/2401/XMake-A-Console-build-system-based-on-XML

vm86 is taken (by us):
http://man7.org/linux/man-pages/man2/vm86.2.html

itrace is taken (by many).

**Comment by waruqi** [comment]
Ok, I will improve the verbose output.

**Comment by waruqi** [comment]
> By the way, you seem to be taking the names
that were already taken.

Yes.

**Comment by stsp** [comment]
Is intermediate transition possible?
For example if I want to only reimplement
configure for now, does xmake support
things like AC_DEFINE, AC_SUBST etc?
Can you just briefly take a look into our
configure.ac and give an estimation of
the porting efforts?

**Comment by stsp** [comment]
autoconf also plays tricks with glibc.
For example we have `AC_SYS_LARGEFILE`
in our configure.ac not to check for
the large-file support, but to actually
enable it! I don't expect xmake to have
something like this, or can it also control
different glibc options?

**Comment by waruqi** [comment]
> AC_DEFINE

You can add macro definition.

```lua
add_defines("HAVE_XXX")
```

Or we can also add macro definition dynamically.

```lua
target("test")
    add_defines("HAVE_XXX")
    after_build(function (target)
         import("core.tool.compiler")
         if compiler.has_features("c_static_assert") then
             target:add("defines", "HAVE_C_STATIC_ASSERT")
         end
    end)
```

> AC_SUBST

define vairable? You can define a lua variable. 

```lua
local var1 = "xxxxx"
```

**Comment by waruqi** [comment]
> autoconf also plays tricks with glibc.
For example we have AC_SYS_LARGEFILE
in our configure.ac not to check for
the large-file support, but to actually
enable it! 

xmake does not add any macro definitions by default. You can add and control it manually.

```lua
add_defines("_GNU_SOURCE=1", "_FILE_OFFSET_BITS=64", "_LARGEFILE_SOURCE")
```

**Comment by stsp** [comment]
>    AC_SUBST
> define vairable? You can define a lua variable.

The way configure communicates with
make, is by generating the files from
templates. In `Makefile.conf.in` we have
strings:
```
CFLAGS:=@CFLAGS@
ALL_CFLAGS:=@DOSEMU_CFLAGS@ ${CFLAGS}
ASFLAGS:=@ASFLAGS@
```
etc, and we get `Makefile.conf` which
has all the `@VARIABLE@` expanded
according to `AC_SUBST`.
This is not only for build, btw.
We also pass an RPM spec file and
man pages via this to put in the version
numbers in. So `AC_SUBST` looks
useful.

> xmake does not add any macro definitions by default. You can add and control it manually.

But then its not portable.
autoconf provides the portable recipes.

**Comment by waruqi** [comment]
You can define an option to pass a command argument to variable in `xmake.lua`

```lua
option("var")
    set_default("default_val")

target("test")
    add_defines("$(var)")
```

```console
$ xmake f --var=xxx
$ xmake
```

We can also pass it to pack rpm if we implement the plugin `xmake pack` in the future.

```console
$ xmake f --var=xxx
$ xmake pack -k rpm
```

But xmake does not pass variable to makefile or makefile.conf because it will build project directly by default.

**Comment by waruqi** [comment]
> But then its not portable.
autoconf provides the portable recipes.

Xmake currently has no way to support the all API features of other build tools, but it provides a flexible, custom extension that allows users to encapsulate and configure by lua function, user modules, etc.

**Comment by stsp** [comment]
> But xmake does not pass variable to makefile or makefile.conf
> because it will build project directly by default.

But you can think of this as building itself.
I.e. we _build_ `dosemu.1` (man page) from
`dosemu.1.in` by substituting `@VARIABLES@`.
In general, to build `file` from `file.in` you need
to substitute the `@VARIABLES@`. So its
nothing but a build rule, why not to support it
if autoconf did?

**Comment by stsp** [comment]
> xmake currently has no way to support the all API features of other build tools,

The thing is, musl is becoming very popular
and some distros are taking it by default. is
providing a portability macros that cover at
least the main libc implementations, is among
your todo goals? We really don't need many,
just to enable some extensions in a glibc/musl
agnostic way.

**Comment by waruqi** [comment]
> But you can think of this as building process.
I.e. we build dosemu.1 (man page) from
dosemu.1.in by substituting @VARIABLES@.
In general, to build file from file.in you need
to substitute the @VARIABLES@. So its
nothing but a build rule, why not to support it
if autoconf did?

This is not necessary, you can use the Lua script directly to achieve, do not need a few lines of code.

```lua
target("test")
    after_install(function (target)
         local vars = {}
         vars.PACKAGE_VERSION = "1.0.0"
         vars.VARIABLES = "xxx"
         local install_man = function (srcfile, dstfile, vars)
             os.cp(srcfile, dstfile)
             io.gsub(dstfile, "@(.+)@", function (name) return vars[name] end)
         end
         install_man("dosemu.1.in", "/xxx/dosemu.1")  
         install_man("xxx.1.in", "/xxx/xxx.1")
    end)
```

You can also wrap the `install_man` script as a custom module and import and use it.

**Comment by stsp** [comment]
We do that on a build stage, not install stage.
If quite common rule can be created for build
stage, this will allow to embed things into our
current build system, which is 1000+ times
more important than just to build man page. :)
Because via that rule I will build `Makefile.conf`.

**Comment by waruqi** [comment]
> The thing is, musl is becoming very popular
and some distros are taking it by default. Is
providing a portability macros that cover at
least the main libc implementations, is among
your todo goals? We really don't need many,
just to enable some extensions in a glibc/musl
agnostic way.

libc has a unified and standard API, so we just need to implement a subset. But there's a lot of difference of features and apis about build-systems.

**Comment by waruqi** [comment]
> We do that on a build stage, not install stage.

You need only use `before_build` or `after_build` instead of `after_install`. I'm just trying to give you an example.

**Comment by stsp** [comment]
> libc has a unified and standard API, so we just need to implement a subset.

What we need is to ask the build system to
check for the extensions we use, and, if they
are present in current libc - enable them.

> You need only use before_build or after_build instead of after_install.
> I'm just trying to give you an example.

But I am not sure I understand why it is before
or after. Its just one of many build rules, like
the rule of building .o from .c. I want to add
the '.in' files into the list of sources rather than
to special-case them. And then to have (or write)
some plugin to xmake that will know how to
proceed that rule.

**Comment by waruqi** [comment]
> But I am not sure I understand why it is before
or after

config -> `before_build` -> `on_build` (build source files) -> `after_build` -> install ..

> Its just one of many build rules, like
the rule of building .o from .c. I want to add
the '.in' files into the list of sources rather than
to special-case them.

Adding `*.in` files directly to the source file list is still not supported now. You can only handle them specially in custom scripts (`before_xxx`, `after_xxx`, etc).

**Comment by waruqi** [comment]
> What we need is to ask the build system to
check for the extensions we use, and, if they
are present in current libc - enable them.

This is similar to tbox API detection ([check_interfaces](https://github.com/tboox/tbox/blob/master/src/xmake.lua)). If some apis present, tbox will enable some features.

**Comment by stsp** [comment]
> Adding *.in files directly to the source file list is still not supported now.

That's fine, but in `make` I could add the custom
rules easily. Is it the case here? Can I add the
compilation rules that will allow me to add new
file types to the sources list?

**Comment by stsp** [comment]
> This is similar to tbox API detection (check_interfaces).
> If some apis present, tbox will enable some features.

No, its a bit different.
Say we need a large file support.
We tell that to the build system.
If its not there - error.
If it is there - build system should
look what libc is here, and enable it
the libc-specific way.
If we just add the glibc-specific flags
you suggested above, then:
1. We don't check if these flags worked
2. We are screwed on non-glibc
So the build system finds what libc is
that, add the needed flags and makes
sure that worked by compiling the test.

**Comment by waruqi** [comment]
It is not supported now, but I will implement this feature to custom compilation rules for the file type in the future.

**Comment by stsp** [comment]
> It is not supported now, but I will implement this
> feature to custom compilation rules for the file type in the future.

OK, so either that, or, preferably, supporting the
`.in` files directly (with AC_SUBST-like API) or both
will get us much closer to the ability of replacing
only the configure system, leaving the build-system
as-is.

**Comment by PaddyMac** [comment]
I've been loosely following this bug because I've been wanting to use DosEmu2 on FreeBSD, and the biggest hurdle has been all of the issues with the configuration step. If you end up not going with a cross-platform build system, is there any way I can help with getting the build system to play nicely with *BSD? I think the main issue is that that A) the name of some programs is different because GNU tools are prefixed with the letter "g" (like gtouch instead of touch or gmake instead of make), and B) the ibraries and headers and so on are located in different paths on the filesystem.

**Comment by stsp** [comment]
In what way autotools is not a cross-platform build?
Could you please show exact example of the failures?
Switching the build system, if ever happens, will be
done _not_ to enrich the amount of target systems,
but just to decrease the build system maintenance
cost. So I would suggest you to open another ticket,
it is really a different problem.

**Comment by waruqi** [comment]
> But you can think of this as building process.
I.e. we build dosemu.1 (man page) from
dosemu.1.in by substituting @VARIABLES@.
In general, to build file from file.in you need
to substitute the @VARIABLES@. So its
nothing but a build rule, why not to support it
if autoconf did?

> But I am not sure I understand why it is before
or after. Its just one of many build rules, like
the rule of building .o from .c. I want to add
the '.in' files into the list of sources rather than
to special-case them. And then to have (or write)
some plugin to xmake that will know how to
proceed that rule.

You can use the custom rule in xmake.lua now (newest version on master), (see [#149](https://github.com/tboox/xmake/issues/149))

```lua
rule("markdown")
    set_extensions(".in")
    on_build(function (target, sourcefile)
        local vars = {PACKAGE_VERSION = "1.0.0", VARIABLES = "xxx"}
        local dstfile = path.join(target:targetdir(), path.basename(sourcefile))
        os.cp(sourcefile, dstfile)
        io.gsub(dstfile, "@(.+)@", function (name) return vars[name] end)
    end)

target("test")
    set_kind("binary")

    -- add rules
    add_rules("man")

    -- add files
    add_files("src/*.c") 
    add_files("src/man/*.in")
```

You can also wrap the `install_man` script as a custom module and import and use it.

**Comment by stsp** [comment]
Yes, looks good.
Are you planning to add the pre-defined
modules for the common things like man?

**Comment by waruqi** [comment]
I have supported the template configuration file in dev branch. like cmake/configure_file and AC_SUBST.

You can see https://github.com/tboox/xmake/issues/320

**Comment by stsp** [comment]
Thanks for an update!
Why `add_config_files("config.h")`
and `add_config_files("config.h.in")`
means entirely the same thing, is
it intentional?
So it seems like it would be possible
to port configure, leaving make system
as is for now, interesting.

**Comment by stsp** [comment]
In fact, I am already advocating
xmake among friends, and am trying
to collect the feed-back. Porting is
a big task, and seeing someone
else doing that, should motivate. :)
Is menuconfig you get for free there,
or one should do some extra coding?

**Comment by stsp** [comment]
Hi @waruqi .
@jschwartzenberg have actually played
around xmake a bit. I cloned his repo and
found that xmake is not in fedora.
Its probably not in ubuntu as well.
Any plans to address that?

**Comment by stsp** [comment]
Also, trying to compile the latest git:
```
link libluajitr.a
make lua-cjson
make[3]: *** Нет правила для сборки цели «lua-cjson/dtoa.c», требуемой для «lua-cjson/dtoa.o».  Останов.
```

And if I run `make build` again, it starts
building from scratches, rather than to
restart from the same place, and to error
out again immediately.

**Comment by waruqi** [comment]
> And if I run `make build` again, it starts
> building from scratches, rather than to
> restart from the same place, and to error
> out again immediately.

You need run `git submodule update --init` to pull all submodule codes first before run `make build`.

Or you can run `bash <(curl -fsSL https://xmake.io/shget.text)` command to install it.

And you can also download https://github.com/xmake-io/xmake/releases/download/v2.3.6/xmake-v2.3.6.gz.run and run it manually.

**Comment by jschwartzenberg** [comment]
@waruqi xmake looks really nice! But it will be rough to depend on it if it's not packaged in distributions as we would like DOSEMU2 to be packaged in the end as well. Right now it only seems to be available through AUR. I searched here: https://pkgs.org/search/?q=xmake

Are you looking into submitting it for inclusion or at least creating a PPA/COPR/etc.?

**Comment by waruqi** [comment]
> Are you looking into submitting it for inclusion or at least creating a PPA/COPR/etc.?

Yes, in the Linux distribution, there is currently only the aur package, but now it has also been submitted to some package management repositories, such as: winget, scoop, termux/pkg, msys2/pacman, homebrew.

In the future I will try to create ppa and more packages.

**Comment by jschwartzenberg** [comment]
> In the future I will try to create ppa and more packages.

That would be great! This is also how DOSEMU2 is currently made available. Hopefully the xmake packages can also flow into the official distribution repositories. Without this we would not be able to adopt it. Thanks!!

**Comment by waruqi** [comment]
> > In the future I will try to create ppa and more packages.
> 
> That would be great! This is also how DOSEMU2 is currently made available. Hopefully the xmake packages can also flow into the official distribution repositories. Without this we would not be able to adopt it. Thanks!!

I have create and submit ubuntu/ppa package. In the future I will create and submit more packages to other linux distributions.

You can see https://github.com/xmake-io/xmake/issues/941

**Comment by stsp** [comment]
Thanks!
copr would be cool to see too.
And thats all we care about.

**Comment by stsp** [comment]
I'll try to summarize our main problem
with `make` and formulate a question
to @waruqi about how xmake assists
in solving it.

TL/DR.
So, while autoconf+make (w/o automake)
has a set of its own complications, and
xmake has a very testy features, our really
big problem comes from the power of a
recursive build. We take for granted that
we can run `make` in any subdirectory,
for example in any plugin directory, and
it will correctly build that part of the project.
In case of a plugin, that would actually
be a ready-to-use solib, which you can
test right away with any previously built
dosemu binary, so no need to re-build
everything else.
Now this all sounds fine and simple,
until you realize that the sub-project
build may depend on an external
resources, like the headers generated
in a higher level directories!
Now the really simple, and CORRECT
solution, is just to fail the partial build
if the outer-level resources are not yet
built or outdated. Then user will re-run
the top-level make to update shared
resources, and you can again do the
partial builds.
Unfortunately, getting used to bad
practices is too easy, and I decided
that the sub-project build failures are
not tolerated. So I started to include
the top-level makefiles into a bottom-level
ones, so that the `make` started in a
sub-directory, can rebuild all of its outer
pre-requisites! I had no idea back then
what consequences will that have. All I
had, was a build failure needed to be
fixed, so I "fixed".
So what are the consequences?
Well, the impossibility of a parallel build.
If any sub-directory can start to rebuild
the outer resources, then the top-level
parallel build will come to a races.
Of course I didn't realize the problem even
then, and started to apply various work-arounds
intuitively. Basically I tried 2 types of work-arounds:
1. have a list of shared resources, and put them
as a pre-requisites of the bottom-level
resources in a top-level makefiles. This is in
addition to also having them as a pre-requisites
individually in the bottom-level makefiles
themselves, so that the make from a sub-dir
can still be used.
2. build same shared resource as a
different files, depending on who's in
need for it. This is an absolutely crazy
work-around... we have bios.s built
multiple times under different names.

So, that's the problem.
What can xmake offer to solve it?
Just not allowing a sub-directory builds?
Or having these builds to fail if the
top-level resource they depend on,
needs to be re-built?
Or, maybe, xmake is magic, and can
automatically gather the list of such
shared resources from the bottom-level
subdirs, and move them on top when
the top-level parallel build is running?
I understand this is too much to desire,
and basically our usage practice is "broken".
But we got use so much to the partial
builds, that I'd like to keep such usage.
If xmake doesn't help with that problem,
then there is no other fundamental
problem that we can't solve with make.
There are many deficiencies, but the
fundamental problem, or at least the
one that requires the non-trivial work-arounds,
is this one.

**Comment by stsp** [comment]
I.e. what I expect from xmake in that
situation, is so that it gather the full
dependency graph. Then look at what
sub-dir it was started, and, based on
these 2 things, arrange the build in a
way that all targets that are a dependency
of several targets (shared dependencies)
are built first. Then the remaining targets
are built in parallel, without the risk of them
to update any shared target.
Such technique will probably finally lead
us out of the parallel build hell we are
currently in.

@waruqi could you please comment
on an xmake's take on that?

**Comment by waruqi** [comment]
We can define some targets for building each target program/library and use `add_deps()` in `target()` to construct a dependency graph. xmake will analyze all the dependencies in the targets, and then perform parallel builds as much as possible.

for example:

```
test4 -> test2 -> test1
          -> test3
```

```lua
target("test1")
    set_kind("static")
    add_files("src/*.c")

target("test2")
    set_kind("binary")
    add_files("src/*.c")
    add_deps("test1")

target("test3")
    set_kind("binary")
    add_files("src/*.c")

target("test4")
    set_kind("binary")
    add_files("src/*.c")
    add_deps("test2", "test3")
```

Then we can run `xmake` to build all default targets. 

And run `xmake build test2` to build the given sub-targets and it's dependences. xmake will build test1 first and build test2.

If the dependent `test1` target has been compiled and it has not any changes, it will not be recompiled unless you execute `xmake --rebuild test2`.

We don't need to enter the corresponding project subdirectory to build the target, all operations can be performed under the project root directory. Of course, executing any xmake command in a subdirectory will have the same effect.

```console
cd projectdir
# build all targets
xmake
# build the given sub-targets (we can define them in subdir/xmake.lua and call includes("subdir") to contain it.)
xmake build xxx
```

In addition, the source files in the target program will be built as parallel as possible. Of course, we can also control the strategy of parallel compilation between targets through configuration

**Comment by stsp** [comment]
Suppose we have this:
```
target("test1")
    set_kind("static")
    add_files("src/*.c")

target("test2")
    set_kind("binary")
    add_files("src/*.c")
    add_deps("test1")

target("test3")
    set_kind("binary")
    add_files("src/*.c")
    add_deps("test1")
```

Here, test1 is a shared target, because
both test2 and test3 depend on it.
If I go to test2 dir and run xmake, it
will build test1 then test2.
If I go to test3 dir, xmake will build
test1 then test3.
That's simple.
But if I go to the top-dir and start
the parallel build from there, how
does xmake make sure test1 is built
only once?
With `make` we have the following:
- starts building test2 and test3 in
parallel
- test2's build process see that test1
needs to be updated and starts updating
it
- test3's build process see that test1
needs to be updated and starts updating
it
- they both corrupt each other's results,
as you can't update test1 by 2 processes
in parallel

What does xmake do?

**Comment by waruqi** [comment]
> If I go to test2 dir and run xmake, it
> will build test1 then test2.
> If I go to test3 dir, xmake will build
> test1 then test3.

We don't need to switch the working directory. Entering a subdirectory has the same effect as any xmake command executed in the root directory. We only need to specify the name of the sub-target program defined in the sub-directory to `xmake build`.

run `xmake build test2` in root directory , it will build test1 then test2.
run `xmake build test3` in root directory , it will build test1 then test3.

> 
> But if I go to the top-dir and start
> the parallel build from there, how
> does xmake make sure test1 is built
> only once?

1. xmake will first construct the dependency graphs of the source files and program files of all the targets that need to be built.

2. Then, xmake will build all c/c++ source files of test1, test2, and test3 in parallel at the same time. By default, xmake assumes that all source files have no dependencies. Of course, this default policy can also be changed.

In addition, xmake only compiles the source files that have actually changed, and the source files whose dependent header files are changed.

3. Finally, xmake first links the test1 dynamic library, and then links the test2/test3 executable program in parallel.

**Comment by stsp** [comment]
OK, that sounds cool, thanks!
I suppose, actually all modern build
systems produce a dependency graph
first. Only with `make` you can recurse
down to the same target in multiple
ways, getting races.
Or is it not the case, and xmake is
better in that regard than some other
build system? Have you happened to
compare with what others do?

**Comment by waruqi** [comment]
I haven't compared it with other build systems, but I think most modern build systems build dependency graphs in similar ways, but there may be some subtle differences.

If there is a better way to deal with it, I will also try to improve it.

**Comment by jschwartzenberg** [comment]
Yes, this sounds similar to Maven's `-pl` switch in combination with its `-am` switch. I would be surprised if a build system would not build up a dependency graph :)

**Comment by stsp** [comment]
> If there is a better way to deal with it, I will also try to improve it.

Well, wikipedia says
`To achieve more accurate dependency graph analysis, Bazel uses content digests rather than file-based timestamps.`
Not sure yet why digests are so much better.
Also here is some very interesting study:
https://www.microsoft.com/en-us/research/uploads/prod/2018/03/build-systems.pdf
Just reading it now. ;)

> I would be surprised if a build system would not build up a dependency graph :)

But with recursive make setup, this is
what you get: no dependency graph.
If you describe the entire build in a
non-recursive manner (in a single
top-level makefile), then you code up
the dependency graph there by hands,
but then the subdir builds do not work.
So with make, things are really really
bad. And I don't know how to make
them better but to switch the build
system.

**Comment by jschwartzenberg** [comment]
Let me correct myself: "I would be surprised if a ***modern*** build system would not build up a dependency graph :)"

**Comment by stsp** [comment]
OK, so that problem with recursive
make setup, was known 20+ years ago:
http://www.real-linux.org.uk/recursivemake.pdf
But writing the non-recursive makefile
for the project as large as today's average
projects are (as dosemu2 also is) - is
basically impossible.
So I guess the compelling case to abandon
make, is found. Before today, I wasn't sure
if make is really not adequate for our needs,
or just I can't properly use it. Now it seems
quite clear that the parallel build hell we have,
is unfixable as long as we stay with make.

Speaking of fdpp, it uses non-recursive
make setup. So it have no problems with
the parallel build, whatsoever. But recently
I converted it to TINY memory model, which
allows a great variety of build tools to be
used instead of nasm+ld.bfd, which we
had to use before. So now fdpp is in a bad
need of a configure step. So again, not
having any problems with make, but we
need to switch to xmake also there, as
xmake provides the configure step for free.

**Comment by stsp** [comment]
@waruqi how does the out-of-tree build
work? In your doc I only found this:
```
Build Project

$ xmake
```
OK, but for out-of-tree build definitely
a different procedure must exist?

**Comment by andrewbird** [comment]
> http://www.real-linux.org.uk/recursivemake.pdf

Indeed, it's an interesting read!

**Comment by waruqi** [comment]
out-of-tree build? we can pass `-P projectdir` to build the given project in outside.

```
xmake -P projectdir
xmake build -P projectdir test1
```

Or we can set `$XMAKE_PROJECT_DIR` 

```
export XMAKE_PROJECT_DIR=projectdir
xmake
xmake build test1
```

**Comment by stsp** [comment]
`$XMAKE_PROJECT_DIR` looks a bit
harsh, as if someone exports it for his
project, it may remain exported, causing
problems.
But -P should do, thanks!

**Comment by stsp** [comment]
I should probably tell it here,
as @waruqi is also here. :)
These 5 years on meson was
a pure agony. Initially I was
very ambitious: sent patches
to meson, reported bugs, etc.
During these days, there appeared
the meson build for thunk_gen
and fdpp. thunk_gen is trivial,
and in fdpp I invested a lot, and
got it to a more-or-less functional
state only when started to call
make from meson rules. But
at least back then I had the
power to define convenient
APIs between meson and make.
But trying to add meson to
comcom64 was a complete
failure. I managed to create a
somewhat working prototype,
but it was breaking with every
dj64 change. Eventually I tried
to at least wrap the existing
buildsystem with meson, but
even that didn't work, so again
I had to post patches to meson.
They are still there unprocessed,
and likely never will, because
the last commit to meson is dated
2 days ago, and 574 opened
pull requests. So I'd say, the
development there have stalled.
They also can't push any new
features to the authors of
ninja-build, who ignore their
requests.

So I really can't predict the future,
but I am 60% confident meson
will not be the supplementary
build system for dosemu2, and
95% confident it won't be its
primary build system.
I really like its concepts and ideas,
but they do not work... 5 years
ago its development was very
active, so I could at least push
my patches forward, but now it
really looks like a dead horse.

Also github now shows the equal
(really! almost identically equal)
amount of commits to xmake and
meson, while 5 years ago the
difference was like in 10 times.

**Comment by waruqi** [comment]
Well, if you have a new project and want to try xmake again, we welcome you at any time. Although we don't have much time, we will still give priority to handling PR and resolving issues as much as possible.

**Comment by stsp** [comment]
Thanks.
@waruqi I have 2 questions:
1. Have something interesting
    appeared in xmake since?
    If so - where to read?
    IIRC back then it didn't even
    support intermediate targets,
    although I vaguely remember.
2. Have you ever considered
    meson adapter for xmake?

No, not kidding about an adapter.
See here for an example:
https://android.googlesource.com/platform/external/mesa3d/+/refs/heads/main/meson_to_hermetic/README.md
So they obviously managed to
parse meson DSL and provide
it with the proxy level of
custom-coded meson objects.
Same technique that I used in
fdpp to compile freedos sources
on toolchains that have no idea
about DOSish C dialects.

Also meson has the fundamental
limitations below:
- no support for "mixed" compilation,
  i.e. cross and native targets within
  the same project. Not even subprojects
  help with that. In cross mode you
  can only build host apps for building
  target apps. But you can't eg install
  them.
- cross-compilation is limited to
  hard-coding the cross-tool names
  into a special text file, verbatim
  (w/o full path fortunately though).
  To me this basically means no
  cross-compilation support at all.
  And devs explicitly say this will
  never change.
- no notion of a preprocessor. So
  some project can't provide some
  "recipe library" for other projects
  to build on top of its recipes. Even
  the mere "include" could already
  make things simpler, but there is
  none.
- no support for static libs, even if
  the patch from one of their core
  devs is in PRs for decades.

And the rest is simply buggy or
partially-implemented functionality.

**Comment by waruqi** [comment]
> Have something interesting
appeared in xmake since?
If so - where to read?
IIRC back then it didn't even
support intermediate targets,
although I vaguely remember.

A lot of things have been improved and a lot of new features have been added. I'm not sure which ones you are interested in. You can check the documentation or changelog.

https://github.com/xmake-io/xmake/blob/dev/CHANGELOG.md

> Have you ever considered
meson adapter for xmake?

Generate meson.build file? 

Maybe I'll consider it in the future, but it's not a priority at the moment and I don't have the time to implement it.

**Comment by stsp** [comment]
> Generate meson.build file?

No, I mean backwards.
Generate xmake.lua from
meson.build, to help transitioning,
or maybe even (considering
the meson2hermetic above)
handle meson.build at runtime
and emulate it with xmake. :)

**Comment by stsp** [comment]
> I'm not sure which ones you are interested in.

Adoption, for example.
Debian?
Fedora?
Termux?
Buildroot?
Flathub?
Homebrew?

**Comment by waruqi** [comment]
> Adoption, for example.
> Debian?
> Fedora?
> Termux?
> Buildroot?
> Flathub?
> Homebrew?

this? https://github.com/xmake-io/xmake/issues/941

- [x] Debian
- [x] Fedora
- [x] Termux
- [x] Homebrew

[![Packaging status](https://repology.org/badge/vertical-allrepos/xmake-io.svg)](https://repology.org/project/xmake-io/versions)

**Comment by waruqi** [comment]
> > Generate meson.build file?
> 
> No, I mean backwards. Generate xmake.lua from meson.build, to help transitioning, or maybe even (considering the meson2hermetic above) handle meson.build at runtime and emulate it with xmake. :)

Not considered, but you can use llm to translate them. e.g chatgpt.

**Comment by stsp** [comment]
Ok, IIRC missing fedora
packaging was a big obstacle,
but it seems resolved.

Now, the very important
requirement is to generate
code during build. This
require many complex rules
with multi-step pipeline with
many intermediate targets.
Will this now work?

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake/tree/dev/tests/projects/other/autogen
https://github.com/xmake-io/xmake/blob/dev/tests/projects/other/autogen/autogen_codedep/xmake.lua
